### PR TITLE
Add Multi-Dimensional Address Generation to Object Fifo

### PR DESCRIPTION
**Pull request by andrej**
I recently added support for the n-dimensional data layout transformations (strides/wraps) of AIE2 to the `dmaBdOp`, #547. This PR adds the same functionality at the objectFifo level.

You can specify two data layout transformations; `toStream` and `fromStream`. `toStream` is applied at the producer side; data in the buffer referenced by the objectFifo is put in the stream by following the strides and wraps given. `fromStream` is the analog on the consumer side; what comes in from the stream is written into the memory addresses obtained by applying the given strides wraps.

Since these data layout transformations require a DMA to perform them, if either `toStream` or `fromStream` is used, a DMA connection between tiles is forced, even if they are adjacent and have shared memory. (This required some light refactoring of the objectFifo code.)

To be more consistent with the `AIEX.dma_nd_memcpy` operation, which first lists lengths and then strides, I also refactored the strides and wraps tuples in both `dmaBdOp` and `objectFifo` (this PR) to now be in order `<wrap, stepsize>` (== synonymously `<length, stride>`).

### CODE REVIEW

File: include/aie/Dialect/AIE/IR/AIE.td [hunk 1]
```diff
@@ -1403,11 +1403,16 @@ def AIE_ObjectFifoCreateOp: AIE_Op<"objectFifo", [HasParent<"DeviceOp">, Symbol]
...
+        OptionalAttr<AIE_DimTupleArrayAttr>:$dimensionsToStream,
+        OptionalAttr<AIE_DimTupleArrayAttr>:$dimensionsFromStream
   );
 
   let assemblyFormat = [{
-    $sym_name `(` $producerTile `,` `{` $consumerTiles `}` `,` $elemNumber `)` attr-dict `:` $elem_type
+    $sym_name `(` $producerTile `,` `{` $consumerTiles `}` `,` $elemNumber 
+      (`,` `toStream` $dimensionsToStream^)?
+      (`,` `fromStream` $dimensionsFromStream^)?
+    `)` attr-dict `:` $elem_type
```
Comment (fifield) [inline include/aie/Dialect/AIE/IR/AIE.td:1415]: Can you put some description of the feature/syntax in the operator's documentation string? And maybe an overview in the PR comments?

Comment (andrej) [reply include/aie/Dialect/AIE/IR/AIE.td:1415]: Yes I will do so and talk about this in the data movement meeting. Just a draft for now

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp [hunk 2]
```diff
@@ -1120,42 +1167,38 @@ struct AIEObjectFifoStatefulTransformPass
...
 
-    for (auto createOp : device.getOps<ObjectFifoCreateOp>()) {
-      objectFifoTiles.insert(createOp.getProducerTileOp());
-      bool shared = false;
+    //===------------------------------------------------------------------===//
+    // Split objectFifos into a consumer end and producer end if needed
+    //===------------------------------------------------------------------===//
+    // We are going to create additional createObjectFifoOps, so get a copy of
+    // all "original" ones before the loop to avoid looping over newly created
+    // ones.
```
Comment (AndraBisca) [inline lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1178]: I apologize if I'm misunderstanding again... I'm having a bit of trouble understanding the two for loops over createOFOps.

Since we create new objectFifos, we do want to loop over them again right? To ensure we create their elements too.

The way I'm reading this rn is: 
- loop once over all ops, find split opportunities and create new OFs,
- loop again over all ops and create their elements (so I don't see the use of createFifoOps).

Couldn't you instead:
- only loop once, any newly create objectFifos will be caught as well,
- for each OF, check if it requiresDMA (either it does because it has data transformation info or it is a broadcast, or it does not require a DMA because it has shared memory and no stream attributes + the bool 'shared' is equal to the result of requiresDMA),
- based on what happened above, create new objectFifos and do all the replacing needed in other ops,
- after you are done iterating over consumers, you can create the objectFifo elements for this objectFifo.

**Review by AndraBisca — APPROVED** [review]
Thank you for taking the time to make these changes and improve the pass!

I am aware this PR is a draft so please feel free to ignore any of my comments that you have already addressed or planned to! I really wanted to give this a look over and this was the only time I could, so I might have done it at a still premature state. 

I think bar the few things I mentioned it looks good already! We can talk about anything I'm still unclear about next week! :)

**Comment by andrej** [comment]
Thanks for looking at this on your vacation! Don't forget to also take some time off!

As for omitting the braces for the one-line if-statements: I personally really dislike that C/C++ lets you do that, so I try to never use that "feature". At best, it saves me less than half a second of typing. At worst, it may cause a novice (or even unfocused/tired expert) later looking at my code minutes trying to debug something after they added a line. (Especially when you have nested {} blocks inside blocks without braces it can get ugly imo.) I know that omitted braces should be obvious to anyone who's used C/C++ for more than a day, but why add the "complication" of omitted braces? It costs almost nothing typing and helps clarity. I'd much rather use two additional characters and not have to worry about it when later I want to add more lines, try to debug something, etc. /rant

That being said, if that violates the style of the rest of this repository (or you'd like to keep the objectFifo code consistent -- this is your baby) I am happy to set aside my preferences for my personal projects and adjust the code accordingly.

Let me address the other comments as replies.

Thanks again for looking at it so quickly.

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp [hunk 3]
```diff
@@ -208,6 +208,37 @@ struct AIEObjectFifoStatefulTransformPass
     return leftShared || rightShared;
   }
 
+  // Return true if an objectFifo between createOp and consumerTileOp requries
+  // a DMA to be set up. This is the case if the tiles are not adjacent (no
+  // shared memory), or if they want to use the multi-dimensional address
+  // genreation features of the DMA.
```
[the remark this answers is not in the document]
Comment (andrej) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:214]: good catches

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp [hunk 4]
```diff
@@ -1167,35 +1210,67 @@ struct AIEObjectFifoStatefulTransformPass
...
+    //===------------------------------------------------------------------===//
+    for (auto createOp : device.getOps<ObjectFifoCreateOp>()) {
+      objectFifoTiles.insert(createOp.getProducerTileOp());
+      bool shared = false;
+      int share_direction = 0;
+      for (auto consumerTile : createOp.getConsumerTiles()) {
+        TileOp consumerTileOp = dyn_cast<TileOp>(consumerTile.getDefiningOp());
+        objectFifoTiles.insert(consumerTileOp);
+        if (!requiresDMAs(createOp, consumerTileOp, share_direction)) {
+          shared = true;
```
[the remark this answers is not in the document]
Comment (andrej) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1271]: This mirrors what your code did previously (plus my changes): If there is more than one consumer, `requiresDMAs` returns `true`, we do not enter the block, so `shared=false`. One way to read `!requiresDMAs` (note the `!`) would be `canUseSharedMemory`, which is not the case if we are broadcasting.

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp (hunk 2 above)
Comment (andrej) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1178]: Yes, you capture well what I'm doing currently. For me, the previous implementation took a while to understand because it changed the set of createFifoOps _as we were iterting over that same set_. I thought it would be easier to conceptually split what we are doing into two phases, like you describe, instead of trying to do everything at once:

1. Split the fifos into consumer/producer side as needed.
2. For the new set of fifos (split ones), create all their elements.

The main reason I changed it wasn't for ease of understanding though; it was to avoid infinite loops. This explanation may be a bit wordy so bear with me.

In the previous implementation, we split the objectFifo into two whenever `shared=false`, creating a new objectFifo on the consumer tile (I'm gonna call this the consumer-side objectFifo). One of the next iterations of the loop then picked up the newly created objectFifo, so we could be sure to create the buffers and locks for both ends.**

The reason you did not end up in an infinite loop is, because eventually, when the iteration got to the consumer objectFifo, this would _not_ be split again. This is because the created consumer end of the objectFifo lists the same tile as both producer and consumer, so it is considered memory-adjacent (`shared=true`). To me, it definitely took a while to see why we are not looping infinitely, repeatedly splitting up the consumer objectFifo, until I discovered this.

Now, I had to make a change that allowed `shared=false`, even if tiles were memory-adjacent. That is why I adjusted that check -- we only use shared memory if the data layout should not be transformed.

But now, in this case, `shared` will always be `false`, even for the newly created consumer side, thus leading to that side being repeatedly split up.

So, that is why I opted for the two loops. Avoid infinitely looping, but also ease of understanding. To me, this is clearer, but obviously I am very biased because I wrote it, so of course I would understand it. My reasoning to why it is clearer would be that we are not changing the set we are iterating over. At least for me, this is easier to wrap my head around (things like DFS/BFS where you change the set as you are iterating take more brain processing for me, and why add that burden when we can also express it without that?).

 **_(Side note: I am not fully convinced that the previous algorithm was provably correct; Consider the case if the new objectFifo was inserted somewhere before the current iterator, i.e. we already iterated past it. Then we would not pick it up again. Maybe you can point me to documentation about the iteration/insertion order -- something that guarantees newly Ops will be added to the end of the iterator -- but that's not the main point I'm trying to make.)_

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrej) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1271]: Yes, looking at this code again, I believe that our objectFifo implementation cannot correctly handle an objectFifo with one memory-adjacent and one "far-away" consumer (i.e. one has shared-memory and one requires DMA). Reason for this being this `break` (and also setting `shared=True` if only one of the consumers has shared memory -- it should be if _all_ have shared memory). Maybe this can be addressed in another PR as the issue would have existed before my changes and is orthogonal to them.

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp (hunk 2 above)
Comment (AndraBisca) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1178]: Thank you so much for taking the time to explain this!!

I apologize that I lacked key comments in there that would have made your understanding easier. But like you say, since I wrote it l, it was much easier for me to understand it.

And while I understand how you changed requiresDMA + isSharedMemory, I didn't realize it would create infinite loops. The two loops now definitely make sense.

By the way, I also think it is clearer to do things in steps and have a dedicated loop for each part of the lowering. I just wanted to see what your additional reasoning was and I'm glad I asked since I now understand!

I know what you mean about the objectFifo ends from the POV of this pass implementation, but users have taken advantage of using the objectFifo with same tile as both producer and consumer and it is still a valid objectFifo with its own state to track, so I think it makes sense to treat it the same.

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp [hunk 5]
```diff
@@ -1167,35 +1210,67 @@ struct AIEObjectFifoStatefulTransformPass
...
         }
-        ObjectFifoCreateOp consumerFifo =
-            createObjectFifo(builder, datatype, consumerFifoName, consumerTile,
-                             consumerTile, consumerObjFifoSize);
+        DimTupleArrayAttr emptyDims =
+            DimTupleArrayAttr::get(builder.getContext(), {});
+        DimTupleArrayAttr singletonFromStreamDims = DimTupleArrayAttr::get(
+            builder.getContext(), {consumerDims[consumerIndex]});
+        DimTupleArrayArrayAttr fromStreamDims = DimTupleArrayArrayAttr::get(
+            builder.getContext(), singletonFromStreamDims);
```
[the remark this answers is not in the document]
Comment (andrej) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1221]: yes good call

File: lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndraBisca) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1271]: I agree that a 'continue' would be more appropriate and also solve the "ensuring all consumer tiles are taken into account". 

But it's true that I tested this (broadcast where every consumer has a loop to unroll) and it worked with the 'break' too....

[the remark this answers is not in the document]
Comment (AndraBisca) [reply lib/Dialect/AIE/Transforms/AIEObjectFifoStatefulTransform.cpp:1271]: If you find any errors that are orthogonal to this, you can create issues and assign them to me! That's a lot of help! :)

**Comment by AndraBisca** [comment]
Thank you for answering so quickly!

For the brackets, you can ignore the comments as I did them out of habit honestly. It's true that I personally like that it's a feature I can use, but the coding of any file in the repo will evolve with our individual styles and that's natural!

File: test/objectFifo-stateful-transform/nd_dma_multiple_consumers_AIE2.mlir [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+ AIE.device(xcve2302) {
+    %tile12 = AIE.tile(1, 2)
+    %tile13 = AIE.tile(1, 3)
+    %tile33 = AIE.tile(3, 3)
+    %tile22 = AIE.tile(2, 2)
+    %tile23 = AIE.tile(2, 3)
+
+    // Even if an objectFifo could be implemented in shared memory, as with 
+    // this case between two adjacent tiles, we need to use DMAs if a data
+    // layout transformation with toStream and fromStream was specified.
```
[the remark this answers is not in the document]
Comment (andrej) [reply test/objectFifo-stateful-transform/nd_dma_multiple_consumers_AIE2.mlir:203]: very good point, this comment is actually misleading due to the broadcast. i'll remove it

**Comment by AndraBisca** [comment]
Work will be continued in #664.
